experimental/parser/lex_number.go

```diff
@@ -0,0 +1,165 @@
+// Copyright 2020-2024 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package parser
+
+import (
+	"math"
+	"strconv"
+	"strings"
+	"unicode"
+
+	"github.com/bufbuild/protocompile/experimental/token"
+)
+
+// lexNumber lexes a number starting at the current cursor.
+func lexNumber(l *lexer) token.Token {
+	tok := lexRawNumber(l)
+	digits := tok.Text()
+
+	// Select the correct base we're going to be parsing.
+	var (
+		prefix      string
+		base        byte
+		legacyOctal bool // Whether this is a C-style 0777 octal.
+	)
+	if len(digits) >= 2 {
+		prefix = digits[:2]
+	}
+	switch prefix {
+	case "0b", "0B":
+		digits = digits[2:]
+		base = 2
+	case "0o", "0O":
+		digits = digits[2:]
+		base = 8
+	case "0x", "0X":
+		digits = digits[2:]
+		base = 16
+	default:
+		if strings.HasPrefix(digits, "0") {
+			base = 8
+			legacyOctal = true
+		} else {
+			base = 10
+		}
+	}
+
+	result, ok := parseInt(digits, base)
+	if !ok {
+		// This may be a floating-point number. That is identified by looking
+		// for a period.
+		if strings.ContainsAny(digits, ".Ee") {
+			if legacyOctal {
+				base = 10
+			}
+
+			if base != 10 {
+				// TODO: We should return ErrInvalidBase here but that requires
+				// validating the syntax of the float to distinguish it from
+				// cases where we want tor return ErrInvalidNumber instead.
+				l.Error(ErrInvalidNumber{Token: tok})
+				token.SetValue(tok, math.NaN())
+				return tok
+			}
+
+			// Delete all underscores.
+			filteredDigits := strings.ReplaceAll(digits, "_", "")
+			hasThousands := len(filteredDigits) < len(digits)
+
+			// We want this to overflow to Infinity as needed, which ParseFloat
+			// will do for us. Otherwise it will ties-to-even as expected.
+			// Currently, the spec does not say "ties-to-even", but it says
+			// "nearest value", which everyone says when they mean
+			// "ties-to-even" whether they know it or not.
```

Comment:
This was my bad for not being precise enough in the spec. Fixed here: https://github.com/bufbuild/protobuf.com/pull/67

(With the spec updated, we can probably omit this part of the comment.)

Reply:
Cool. I left some comments on that fix.